src/java.base/share/classes/java/math/BigDecimal.java

```diff
@@ -1797,36 +1797,60 @@ public BigDecimal divide(BigDecimal divisor) {
         if (this.signum() == 0) // 0/y
             return zeroValueOf(preferredScale);
         else {
-            /*
-             * If the quotient this/divisor has a terminating decimal
-             * expansion, the expansion can have no more than
-             * (a.precision() + ceil(10*b.precision)/3) digits.
-             * Therefore, create a MathContext object with this
-             * precision and do a divide with the UNNECESSARY rounding
-             * mode.
+            /* The technique used is the following:
+             * take a/b, compute b' = b/(2^i 5^j), where
+             * i = max{n | b ≡ 0 mod 2^n} and j = max{n | b ≡ 0 mod 5^n}.
+             * If a ≢ 0 mod b', then a/b is not a finite decimal number. Otherwise:
+             *   - if i ≤ j, then a/b = (a/b') ⋅ 2^(j-i) / 10^j;
+             *   - if i > j, then a/b = (a/b') ⋅ 5^(i-j) / 10^i.
              */
-            MathContext mc = new MathContext( (int)Math.min(this.precision() +
-                                                            (long)Math.ceil(10.0*divisor.precision()/3.0),
-                                                            Integer.MAX_VALUE),
-                                              RoundingMode.UNNECESSARY);
-            BigDecimal quotient;
-            try {
-                quotient = this.divide(divisor, mc);
-            } catch (ArithmeticException e) {
-                throw new ArithmeticException("Non-terminating decimal expansion; " +
-                                              "no exact representable decimal result.");
+            BigInteger den = divisor.unscaledValue();
+            int powsOf2 = den.getLowestSetBit();
+            den = den.shiftRight(powsOf2); // Remove powers of 2
+
+            int powsOf5 = 0;
+            // Remove and count powers of 5
+            BigInteger[] qr;
+            int i;
+            for (i = 0; ; i++) {
```

Comment:
What about a comment explaining the idea of this and the following fast loops using 5^(2^i) powers to find the correct `powsOf5` (the `j` in the comment)?